Doc and build related cleanups + slight refactoring, add support for Asciidoctor

jnpkrn opened the pull request.

jnpkrn commented:
The goal is to eventually replace those nasty `sed`s with a single XSLT,
but the current band-aid solution allows us to start shooting for a full
`asciidoctor` (along with `asciidoc`) compatibility early.

Tentatively also:
https://github.com/travis-ci/apt-package-whitelist/issues/4306

jnpkrn commented:
Also tentatively, the plan for the future would be to inverse format
"authoritativness", i.e. switch to using Asciidoctor's proper
(non-compat) and in turn authoritative contemporary AsciiDoc
syntax while providing backward compatibility through the tweaked
configuration file:
https://github.com/asciidoctor/asciidoctor/blob/master/compat/asciidoc.conf

kgaillot commented on doc/Pacemaker_Explained/en-US/Ch-Alerts.txt:
```diff
@@ -75,7 +75,7 @@ to affect how Pacemaker calls them.
 .Meta-Attributes of an Alert
 
-[width="95%",cols="m,1,2<a",options="header",align="center"]
+[width="95%",cols="m,1,<2",options="header",align="center"]
```
Was there a reason to strip all the a's out?

jnpkrn commented:
On 09/08/18 10:14 -0700, Ken Gaillot wrote:
>> @@ -263,6 +263,7 @@ dnl ===============================================
>  dnl General Processing
>  dnl ===============================================
>  
> +AC_PROG_MAKE_SET
> 
> Do we really need this and @SET_MAKE@ to be explicit? In both my 1.1
> and master checkouts, config.log shows that configure sets SET_MAKE,
> and the generated Makefile.in's contain @SET_MAKE@, so it seems to
> be the default behavior.

Good catch, mostly matter of the whole picture scattered amongst
multiple autotools, that when used together, tend to do everything
auto-magically, and I tend to forget about that :-/

AM_INIT_AUTOMAKE macro performs AC_PROG_MAKE_SET equivalent since at
least 1996.

Ditto SET_MAKE variable (since "initial revision"/1994).

Will drop those overly keen compatibility measures.
At least there would be something redundant, which failsafe compared
to missing :-)

>> @@ -565,8 +565,9 @@ if test x"${MANPAGE_XSLT}" != x""; then
>      PCMK_FEATURES="$PCMK_FEATURES agent-manpages"
>  fi
>  
> -AM_CONDITIONAL(BUILD_ASCIIDOC, test x"${ASCIIDOC}" != x"")
> -if test x"${ASCIIDOC}" != x""; then
> +AM_CONDITIONAL([IS_ASCIIDOC], [test "x${ASCIIDOC_CONV}" = xasciidoc])
> 
> ASCIIDOC_CONV will be a full path
> 
>> @@ -57,18 +57,27 @@ endif
>  # articles and only books can have appendices) and then strip out the
>  # bits we don't want/need
>  #
> +# XXX Per-chapter approach seems to preclude cross-chapter references
> 
> Not sure what you mean, we use cross-chapter references, e.g.
> 
>     doc/Pacemaker_Explained/en-US/Ch-Advanced-Options.txt:[[s-failure-handling]]
>     doc/Pacemaker_Explained/en-US/Ch-Options.txt:and +migration-threshold+ (see <<s-failure-handling>>).

TBH, it was my recollection from the past that some kind of validation
failed because of that:

https://github.com/ClusterLabs/pacemaker/pull/1523#issuecomment-399533238

Trying to figure out what the original problem was, I've discovered
that -- indeed -- it was all matter of subtleties in the column
specifier!  See below at relevant place...

>>  %.xml:  %.txt
> -	$(AM_V_ASCII)$(ASCIIDOC) -b docbook -a cli_name=$(ASCIIDOC_CLI_TYPE) -a $(ASCIIDOC_CLI_TYPE)=true -d book -o $@ $<
> +if IS_ASCIIDOC
> +	$(AM_V_ASCII)$(ASCIIDOC_CONV) -b docbook -a cli_name=$(ASCIIDOC_CLI_TYPE) -a $(ASCIIDOC_CLI_TYPE)=true -d book -o $@ $<
> +else
> +	$(AM_V_ASCII)$(ASCIIDOC_CONV) -b docbook45 -a cli_name=$(ASCIIDOC_CLI_TYPE) -a $(ASCIIDOC_CLI_TYPE)=true -d book -o $@ $<
> +endif
> 
> This would be more readable setting a variable for the -b arg in the
> if-else and then having a single command below it, to easily tell
> what's different

The point is that two not necessarily compatible programs are
configured, so the orthogonal tweaks may be needed, so perhaps
better to keep the free way like this.

I'll drop ASCIIDOC_CLI_TYPE references that are a leftover from
81b00e6da9e3bceb315d3b10fc9ab8c84c2ea1bd, though.

>> @@ -75,7 +75,7 @@ to affect how Pacemaker calls them.
>  
>  .Meta-Attributes of an Alert
>  
> -[width="95%",cols="m,1,2<a",options="header",align="center"]
> +[width="95%",cols="m,1,<2",options="header",align="center"]
> 
> Was there a reason to strip all the a's out?

Now, we are getting to something interesting, and related to the
original cross-chapter references per above.

Comparison of the output for this source snippet:

> |value-source
> |+literal+
> | [[s-rules-nodeattrexprs-valuesource]]
> How the +value+ is derived. Allowed values:

With 'a' = asciidoc, reference target is rendered in docbook tags like
this:

> <literallayout class="monospaced"> [[s-rules-nodeattrexprs-valuesource]]
> How the +value+ is derived. Allowed values:</literallayout>

whereas with default (optionally denoted with 'd')

> <para><anchor
> id="s-rules-nodeattrexprs-valuesource"
> xreflabel="[s-rules-nodeattrexprs-valuesource]"/>
> How the <literal>value</literal> is derived. Allowed values:</para>

Apparently, the refences (type IDREF) won't work in the former case
when misinterpreted like that.  Takeaway hence is that implicit
default is saner and it's asctually questionable why anyone would
want to touch on something that seems to work well right away.

-- 
Jan (Poki)

jnpkrn commented on configure.ac:
```diff
@@ -565,8 +565,9 @@ if test x"${MANPAGE_XSLT}" != x""; then
 
-AM_CONDITIONAL(BUILD_ASCIIDOC, test x"${ASCIIDOC}" != x"")
-if test x"${ASCIIDOC}" != x""; then
+AM_CONDITIONAL([IS_ASCIIDOC], [test "x${ASCIIDOC_CONV}" = xasciidoc])
```
Ah, oversight from `AC_CHECK_PROG` confusion. Thanks

kgaillot commented on Makefile.common:
```diff
@@ -57,18 +57,27 @@ endif
-	-grep -vqis "<appendix" $@ && sed -i 's/book>/chapter>/g' $@ # Rename to chapter
+	$(AM_V_at)sed -i 's/
+	$(AM_V_at)sed -i 's/
//' $@                # Fix line endings
```
Are the above two lines a github display issue or a code issue?

jnpkrn replied:
Oops, sorry for this, might be related to me copying something from
garbled output.

jnpkrn replied:
(so yes, root cause might have been a display issue, but rather on my terminal)

jnpkrn replied:
Opened https://github.com/isaacs/github/issues/1329 as a reminder
about this confusing behaviour.

kgaillot commented on doc/Pacemaker_Explained/en-US/Ch-Constraints.txt:
```diff
@@ -250,7 +250,7 @@ alternatively, a group (see <<group-resources>>).
 
 .Properties of a rsc_order Constraint
-[width="95%",cols="1m,1,4<a",options="header",align="center"]
+[width="95%",cols="1m,1,<4",options="header",align="center"]
```
Does this affect the bulleted list in the kind description?

jnpkrn replied:
Yes, it does, and thanks for pointing that out.
It suffices to mark the affected cells as `asciidoc`, though.
Hidden comments ditto.

jnpkrn replied:
It also means that anchors cannot likely be placed into lists withing
the table cells, which may be the most probable reason of placing
the anchor at all :-/

kgaillot commented on lib/lrmd/lrmd_client.c:
```diff
@@ -769,7 +769,7 @@ lrmd_api_is_connected(lrmd_t * lrmd)
  * \param[in]  data          Command data XML to add to the sent command
  * \param[out] output_data   If expecting a reply, it will be stored here
- * \param[in]  timeout       Timeout in milliseconds (if 0, defaults to 1000);
+ * \param[in]  timeout       Timeout in milliseconds (if 0, defaults to 5000);
```
Unfortunately it looks more complicated than this. This function uses the argument twice:

* It is passed to lrmd_create_op(), which adds it to the API command XML as F_LRMD_TIMEOUT, which the executor will parse into ((lrmd_cmd_t*)cmd)->timeout. I don't see anywhere that the executor defaults 0 there to some other value, so it may be left alone, which would likely affect log messages and systemd action retries, and potentially get passed to the stonith API's monitor (which would indeed end up defaulting to 1000) or resources_action_create(), which does who knows what with it.

* It is passed to lrmd_send_xml(), which can pass it to either crm_ipc_send() (which defaults 0 to 5000) or lrmd_tls_send_recv() (which defaults 0 to 10000)

Easiest solution here would be to just leave off the parenthesized bit, or, with a bit of faith, "(if 0, defaults to a reasonable value)".

jnpkrn replied:
True, I'd cast an opposite bias (just as bad), thanks for catching that.

jnpkrn commented:
Changed a bit wrt. control characters -- we really want them out of the
project, hence there's now also a CI check to the effect.  Legacy `LGPLv2.1`
version was also affected (control characters for nicer paging in Emacs,
I believe)

Also fixed another breakage wrt. `Makefile.common`, `%.xml:  %.txt` recipe
(final target needs only be present with the last command).

jnpkrn commented:
Ready to merge if nothing else is found.

jnpkrn commented:
Hm, realized I want to flip the order of `MAINT=1` tests, but there were chained
problems (piping correctly, xargs not returning 0, yaml formatting!), but there are
cleared now.

kgaillot commented:
The control-character check is failing on travis:

https://travis-ci.org/ClusterLabs/pacemaker/jobs/420093844

jnpkrn commented:
On 24/08/18 09:53 -0700, Ken Gaillot wrote:
> The control-character check is failing on travis:
> 
> https://travis-ci.org/ClusterLabs/pacemaker/jobs/420093844

(Luckily) not.

It's caused by an intermittent unavailability of
http://relaxng.org/relaxng.rng URL, for whatever reason:

> error : Operation in progress
> warning: failed to load external entity "http://relaxng.org/relaxng.rng"
> Relax-NG parser error : xmlRelaxNGParse: could not load http://relaxng.org/relaxng.rng
> Relax-NG schema http://relaxng.org/relaxng.rng failed to compile

Actually looks like we accidentally discovered an issue in libxml2,
whereby it is not ready to handle EINPROGRESS or something :-D

I'll think of the solutions for that (or unavailability of that host).

-- 
Nazdar,
Poki

jnpkrn commented:
Oh, and this `MAINT=1` run lasted `4 min 22 sec`, unlike with
times under a minute when it finishes successfully.

jnpkrn commented:
See #1570
